models/src/node/mod.rs

```diff
@@ -0,0 +1,264 @@
+)]
+pub struct BottlerocketNodeSpec {
+    /// Records the desired state of the `BottlerocketNode`
+    pub state: BottlerocketNodeState,
+    /// The time at which the most recent state was set as the desired state.
+    state_transition_timestamp: Option<String>,
+    /// The desired update version, if any.
+    #[validate(regex = "SEMVER_RE")]
+    version: Option<String>,
+}
+
+impl BottlerocketNode {
+    /// Creates a `BottlerocketNodeSelector` from this `BottlerocketNode`.
+    pub fn selector(&self) -> Result<BottlerocketNodeSelector> {
+        BottlerocketNodeSelector::from_bottlerocket_node(self)
+    }
+
+    /// Returns whether or not a node has reached the state requested by its spec.
+    pub fn has_reached_desired_state(&self) -> bool {
+        self.status.as_ref().map_or(false, |node_status| {
+            node_status.current_state == self.spec.state
+        })
+    }
+}
+
+impl BottlerocketNodeSpec {
+    pub fn new(
+        state: BottlerocketNodeState,
+        state_transition_timestamp: Option<DateTime<Utc>>,
+        version: Option<Version>,
+    ) -> Self {
+        let state_transition_timestamp = state_transition_timestamp.map(|ts| ts.to_rfc3339());
+        let version = version.map(|v| v.to_string());
+        BottlerocketNodeSpec {
+            state,
+            state_transition_timestamp,
+            version,
+        }
+    }
+
+    /// Creates a new BottlerocketNodeSpec, using the current time as the timestamp for timeout purposes.
+    pub fn new_starting_now(state: BottlerocketNodeState, version: Option<Version>) -> Self {
+        Self::new(state, Some(Utc::now()), version)
+    }
+
+    /// JsonSchema cannot appropriately handle DateTime objects. This accessor returns the transition timestamp
+    /// as a DateTime.
+    pub fn state_timestamp(&self) -> Result<Option<DateTime<Utc>>> {
+        self.state_transition_timestamp
+            .as_ref()
+            .map(|ts_str| {
+                DateTime::parse_from_rfc3339(ts_str)
+                    // Convert `DateTime<FixedOffset>` into `DateTime<Utc>`
+                    .map(|ts| ts.into())
+                    .context(error::TimestampFormat)
+            })
+            .transpose()
+    }
+
+    /// Returns the desired version for this BottlerocketNode.
+    pub fn version(&self) -> Option<Version> {
+        // We know this won't panic because we have a regex requirement on this attribute, which is enforced by the k8s schema.
+        self.version.as_ref().map(|v| Version::from_str(v).unwrap())
+    }
+}
+
+/// `BottlerocketNodeStatus` surfaces the current state of a bottlerocket node. The status is updated by the host agent,
+/// while the spec is updated by the brupop controller.
+#[derive(Clone, Debug, Default, Serialize, Deserialize, Eq, PartialEq, JsonSchema)]
+pub struct BottlerocketNodeStatus {
+    #[validate(regex = "SEMVER_RE")]
+    current_version: String,
```

Comment:
Why do we store `String` here and convert between String and Version? Why didn't we directly store Version?

Reply:
This `BottlerocketNodeStatus` is being serialized into an object in Kubernetes. For this to work, it needs to support `Serialize`, `Deserialize`, and `JsonSchema`, which I believe isn't all true for `semver::Version`. Kubernetes, however, does allow you to serialize strings and add regular expressions to the schema which ensure they match a pattern. This `validate` macro adds the schema requirement for us.